lib/semantics/io.cc

```diff
@@ -0,0 +1,619 @@
+// Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "io.h"
+#include "expression.h"
+#include "tools.h"
+#include "../evaluate/fold.h"
+#include "../parser/tools.h"
+
+namespace Fortran::semantics {
+
+// TODO: C1234, C1235 -- defined I/O constraints
+
+using DefaultCharConstantType =
+    evaluate::Constant<evaluate::Type<common::TypeCategory::Character, 1>>;
+
+void IoChecker::Enter(const parser::ConnectSpec &spec) {
+  // ConnectSpec context FileNameExpr
+  if (std::get_if<parser::FileNameExpr>(&spec.u)) {
+    SetSpecifier(SpecifierKind::File);
+  }
+}
+
+void IoChecker::Enter(const parser::ConnectSpec::CharExpr &spec) {
+  SpecifierKind specKind{};
+  using ParseKind = parser::ConnectSpec::CharExpr::Kind;
+  switch (std::get<ParseKind>(spec.t)) {
+  case ParseKind::Access: specKind = SpecifierKind::Access; break;
+  case ParseKind::Action: specKind = SpecifierKind::Action; break;
+  case ParseKind::Asynchronous: specKind = SpecifierKind::Asynchronous; break;
+  case ParseKind::Blank: specKind = SpecifierKind::Blank; break;
+  case ParseKind::Decimal: specKind = SpecifierKind::Decimal; break;
+  case ParseKind::Delim: specKind = SpecifierKind::Delim; break;
+  case ParseKind::Encoding: specKind = SpecifierKind::Encoding; break;
+  case ParseKind::Form: specKind = SpecifierKind::Form; break;
+  case ParseKind::Pad: specKind = SpecifierKind::Pad; break;
+  case ParseKind::Position: specKind = SpecifierKind::Position; break;
+  case ParseKind::Round: specKind = SpecifierKind::Round; break;
+  case ParseKind::Sign: specKind = SpecifierKind::Sign; break;
+  case ParseKind::Convert: specKind = SpecifierKind::Convert; break;
+  case ParseKind::Dispose: specKind = SpecifierKind::Dispose; break;
+  }
+  SetSpecifier(specKind);
+  if (const auto *expr{GetExpr(std::get<parser::ScalarDefaultCharExpr>(spec.t)
+                                   .thing.thing.value())}) {
+    const auto &foldExpr{
+        evaluate::Fold(context_.foldingContext(), common::Clone(*expr))};
+    if (const DefaultCharConstantType *
+        charConst{evaluate::UnwrapExpr<DefaultCharConstantType>(foldExpr)}) {
+      std::string s{parser::ToUpperCaseLetters(**charConst)};
+      if (specKind == SpecifierKind::Access) {
+        hasKnownAccess_ = true;
+        hasAccessDirect_ = s == "DIRECT"s;
```

Comment:
Just FYI, you can compare `std::string` to regular old character literal strings if you like, without the `s` suffix.

Reply:
I've removed s suffixes.